jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketCloseTest.java

```diff
@@ -425,6 +428,50 @@ public void testAbnormalCloseStatusIsHardClose(String scheme) throws Exception
         assertThat(CloseStatus.getCloseStatus(frame).getCode(), is(CloseStatus.SERVER_ERROR));
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {WS_SCHEME, WSS_SCHEME})
+    public void doubleClose(String scheme) throws Exception
+    {
+        setup(State.OPEN, scheme);
+
+        CountDownLatchCallback callback1 = new CountDownLatchCallback();
+        server.handler.getCoreSession().close(CloseStatus.SERVER_ERROR, "server error should succeed", callback1);
+        CountDownLatchCallback callback2 = new CountDownLatchCallback();
+        server.handler.getCoreSession().close(CloseStatus.PROTOCOL, "protocol error should fail", callback2);
+
+        assertTrue(callback1.succeeded.await(5, TimeUnit.SECONDS));
+        assertThat(callback2.failed.get(5, TimeUnit.SECONDS), instanceOf(ClosedChannelException.class));
+
+        assertTrue(server.handler.closed.await(5, TimeUnit.SECONDS));
+        assertThat(server.handler.closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(server.handler.closeStatus.getReason(), containsString("server error should succeed"));
+
+        Frame frame = receiveFrame(client.getInputStream());
+        assertThat(CloseStatus.getCloseStatus(frame).getCode(), is(CloseStatus.SERVER_ERROR));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {WS_SCHEME, WSS_SCHEME})
+    public void doubleCloseAbnormalOvertakesNormalClose(String scheme) throws Exception
+    {
+        setup(State.OPEN, scheme);
+
+        CountDownLatchCallback callback1 = new CountDownLatchCallback();
+        server.handler.getCoreSession().close(CloseStatus.NORMAL, "normal close (client does not complete close handshake)", callback1);
+        CountDownLatchCallback callback2 = new CountDownLatchCallback();
+        server.handler.getCoreSession().close(CloseStatus.SERVER_ERROR, "error close should overtake normal close", callback2);
+
+        assertTrue(callback1.succeeded.await(5, TimeUnit.SECONDS));
+        assertThat(callback2.failed.get(5, TimeUnit.SECONDS), instanceOf(ClosedChannelException.class));
+
+        assertTrue(server.handler.closed.await(5, TimeUnit.SECONDS));
+        assertThat(server.handler.closeStatus.getCode(), is(CloseStatus.SERVER_ERROR));
+        assertThat(server.handler.closeStatus.getReason(), containsString("error close should overtake normal close"));
+
+        Frame frame = receiveFrame(client.getInputStream());
+        assertThat(CloseStatus.getCloseStatus(frame).getCode(), is(CloseStatus.NORMAL));
+    }
```

Comment:
As we discussed in the call, currently abnormal status codes are coded to hard close the connection. So if one is sent in the middle of a close handshake it will interrupt the close handshake and the connection will be closed because of the abnormal status code and not because of the successful close handshake response.

The success or failure of the close callback indicates only whether the close frame could be sent, the `onClosed` notification tells you the actual reason for the close which could be different for various reasons.